app/models/container_deployment.rb

```diff
@@ -6,5 +6,387 @@ class ContainerDeployment < ApplicationRecord
+        end
+        container_deployment_node.labels = JSON.parse(labels[vm_attr["vmName"]].to_json).to_h if labels && labels[vm_attr["vmName"]]
+        container_deployment_node.tag_add tags[index]
+        container_deployment_node.save!
+      end
+    end
+  end
+
+  # will be removed once moving to yaml format, no need to review method
+  def add_role(role_name)
+    template = <<eos
+[#{role_name}]
+#{send("#{role_name}_addresses".to_sym).join("\n")}
+eos
+    template
+  end
+
+  def create_deployment(params, user)
+    self.version = "v3"
+    self.kind = params["providerType"].include?("openshiftOrigin") ? "origin" : "openshift-enterprise"
+    self.method_type = params["provisionOn"]
+    tags = create_needed_tags(params)
+    labels = params[:labels]
+    if method_type.include? "existingVms"
+      self.deployed_on_ems = ExtManagementSystem.find params["existingProviderId"]
+    elsif method_type.include? "newVms"
+      ContainerDeployment.add_basic_root_template
+      self.deployed_on_ems = ExtManagementSystem.find params["newVmProviderId"]
+      public_key, private_key = generate_ssh_keys
+      params["deploymentKey"] = private_key
+      params["public_key"] = public_key
+      params["deploymentUsername"] = "root"
+    end
+    deployment_master = params["masters"].shift
+    create_deployment_nodes([params["nodes"], params["masters"], [deployment_master]] + add_additional_roles(params), labels, tags)
+    create_deployment_authentication(params["authentication"])
+    create_deployment_authentication("ssh" => {"userid" => params["deploymentUsername"], "auth_key" => params["deploymentKey"], "public_key" => params["public_key"]}, "mode" => "ssh")
+    create_deployment_authentication("rhsm" => {"userid" => params["rhnUsername"], "password" => params["rhnPassword"], "rhsm_sku" => params["rhnSKU"]}, "mode" => "rhsm")
+    save!
+    create_automation_request(generate_automation_params(params), user)
+  end
+
+  def create_needed_tags(params = {})
+    node_tag = create_or_get_tag("node").name
+    master_tag = create_or_get_tag("master").name
+    deployment_master_tag = create_or_get_tag("deployment_master").name
+    tags = [node_tag, master_tag, deployment_master_tag]
+    nfs_tag = create_or_get_tag("nfs").name if params["nfs_id_or_ip"]
+    tags += [nfs_tag] if nfs_tag
+    tags
+  end
+
+  # will be removed once moving to yaml format, no need to review method
+  def add_additional_roles(params)
+    additional_roles_vms = []
+    additional_roles_vms = [params["nfs_id_or_ip"]] if params["nfs_id_or_ip"]
+    additional_roles_vms
+  end
+
+  def create_or_get_tag(tag)
+    ent = Tag.find_by_name(tag)
+    unless ent
+      ent = Tag.new(:name => tag)
+      ent.save
+    end
+    ent
+  end
+
+  def create_automation_request(parameters, user)
+    uri_parts = {
+      "namespace" => "ManageIQ/Deployment/ContainerProvider/System/StateMachine",
+      "class"     => "Deployment",
+      "instance"  => "Deployment"
```

Comment:
Its related to automation-engine, when you start an automation task it start with a class that has a few state/function that are being called (can be called with different values depends on the instance), the instance is the call and values that those function will be called (a class can have many instances).
The Email class has many instances:
![selection_024](https://cloud.githubusercontent.com/assets/11405684/15565921/b3480028-2326-11e6-9868-7482c5a7bbe5.png)

instance example:
![selection_023](https://cloud.githubusercontent.com/assets/11405684/15565886/7a1cc91e-2326-11e6-84e2-869eb5042c31.png)